feature/Create-Record-transform

ailegion opened the pull request:
added: Create Record transform

JIRA: https://issues.cask.co/browse/PLUGIN-391

ailegion committed:
added: Create Record Transform

rmstar commented on pom.xml:
```diff
@@ -0,0 +1,267 @@
+  <properties>
+    <!-- properties for script build step that creates the config files for the artifacts -->
+    <data.pipeline.parent>system:cdap-data-pipeline[6.1.1,7.0.0-SNAPSHOT)</data.pipeline.parent>
+    <data.stream.parent>system:cdap-data-streams[6.1.1,7.0.0-SNAPSHOT)</data.stream.parent>
```
lower bound should match `cdap.version`

ailegion replied:
fixed

rmstar commented on src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java:
```diff
@@ -0,0 +1,385 @@
+  public void initialize(TransformContext context) throws Exception {
+    super.initialize(context);
+    FailureCollector failureCollector = context.getFailureCollector();
+    createRecordTransformConfig.validate(failureCollector);
```
Add validation in `prepareRun()`. Also emit lineage.

ailegion replied:
added

rmstar commented on docs/CreateRecord-transform.md:
```diff
@@ -0,0 +1,62 @@
+		"id": ["order_id"],
+		"product_id": ["product_id"],
+		"amount": ["amount"],
+		"description": ["order_description"]
```
Please provide an example mapping for the case where the hierarchy is more than 1 level deep. Please add a unit test for this case as well.

ailegion replied:
updated

rmstar reviewed (commented):
Please add screenshots of the plugin properties and documentation widgets.

ailegion commented:
> Please add screenshots of the plugin properties and documentation widgets.

Requested screenshots:
![CreateRecords-Widgets](https://user-images.githubusercontent.com/7079754/96589464-20152700-12e5-11eb-9b87-27a5c7296f42.png)
![CreateTransform-DocumentationPart1](https://user-images.githubusercontent.com/7079754/96589467-21465400-12e5-11eb-8f68-5c66edf57be9.png)
![CreateTransform-DocumentationPart2](https://user-images.githubusercontent.com/7079754/96589469-21deea80-12e5-11eb-865b-98515b9f17cf.png)

rmstar commented on src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java:
```diff
@@ -382,4 +398,49 @@ private Object getField(StructuredRecord structuredRecord, JsonArray pathMap) {
+  private LinkedList<FieldOperation> generateLineageRecord(String recordName, JsonElement record) {
+    LinkedList<FieldOperation> fieldOperations = new LinkedList<>();
+    List<String> inputFields = new ArrayList<>();
+    List<String> recordFields = new ArrayList<>();
```
Is `recordFields` the list of output fields generated by the operation?

ailegion replied:
updated - changed to ``outputFields``

rmstar commented on src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java:
```diff
@@ -382,4 +398,49 @@ private Object getField(StructuredRecord structuredRecord, JsonArray pathMap) {
+    for (Map.Entry<String, JsonElement> recordNode : record.getAsJsonObject().entrySet()) {
+      JsonElement value = recordNode.getValue();
+      if (value.isJsonArray()) {
+        recordFields.add(recordNode.getKey());
```
please add a comment explaining what you're doing here

ailegion replied:
added comment

rmstar commented on src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java:
```diff
@@ -382,4 +398,49 @@ private Object getField(StructuredRecord structuredRecord, JsonArray pathMap) {
+      }
+      if (value.isJsonObject()) {
+        fieldOperations.addAll(generateLineageRecord(recordNode.getKey(), recordNode.getValue()));
+        recordFields.add(recordNode.getKey());
```
we update recordFields but not inputFields - is that correct? Please add a comment here explaining what we're doing here.

ailegion replied:
added comment - when type is JSON object in map config that means we have nested field on output, we call the function recursively in order to map input fields and output fields.

CuriousVini commented:
I think recent merge caused conflicts. Could you please resolve conflicts too along with comments?

ailegion commented:
> I think recent merge caused conflicts. Could you please resolve conflicts too along with comments?

resolved the conflicts and addressed the comments.

rmstar commented on src/main/java/io/cdap/plugin/transform/CreateRecordTransform.java:
```diff
@@ -0,0 +1,448 @@
+      if (value.isJsonObject()) {
+        // the field is nested object - make recursive call to map nested object fields
+        fieldOperations.addAll(generateLineageRecord(recordNode.getKey(), recordNode.getValue()));
+        outputFields.add(recordNode.getKey());
```
We are adding an output field here, but no input field. Is that correct?

For example, if we have input table:
```
| field                 | type         |
| ------------          | ------------ |
| customer_profile      | Record       |
| - customer_name       | String       |
| - customer_phone      | String       |
```

with mapping:
```	
"customer": {
		"name": ["customer_profile", "customer_name"],
		"phone": ["customer_profile", "customer_phone"]
}
```

we would add a FieldTransOperation with inputFields=[], outputFields=["customer"].

ailegion replied:
When it's nested record it calls recursive function (line 436) - which will hit condition on line 429 for fields inside the nested record and will get input and output fields and then create FieldTransformOperation in line 445 for the nested record. 

In the example you gave the "customer_profile" input field will be in the new FieldTransformOperation generated inside the recursive call.

rmstar reviewed (approved):
Please squash commits. Thanks!

ailegion commented:
> Please squash commits. Thanks!

squashed
